### PR TITLE
Emoji Picker

### PR DESCRIPTION
**Pull request by sslater11**
- Added an emoji picker, solves #43 
- Changed the swipes on the number pad key allowing cut copy and paste
- Added undo/redo, solves #284 
- Updated screenshot
- Changed most keyboards to use the new emoji key, replacing settings key. See below for issues.

There's a single git commit for incrementing the indentation by 4 spaces, because I have put an if/else clause in KeyboardScreen.kt. The indentation messed up the git diff and it was hard to see what code I had changed, so I fixed the indentation as a separate commit.
It's worth mentioning, because new PRs might have issues if they're adding commits to the old KeyboardScreen.kt file.

For undo/redo, I tried to make an object KeyActions.Undo, and put the code there, but it just doesn't work, so I have done it like this in the CommonKeys.kt file
`action = KeyAction.SendEvent( KeyEvent(/* downTime= */ 0, /* eventTime= */ 0, KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_Z, /* repeat= */ 0, /* metaState= */ KeyEvent.META_CTRL_ON)
`

In KeyboardScreen.kt I have duplicate code for the massive function call KeyboardKey(). I couldn't figure out a better way to do it.

**Broken Keyboards:**
Keyboards that don't work with the emoji key, and so may be missing the ability to change layouts, because I removed that option from the number pad key and put it on the emoji key.
I don't use these layouts, so I'm unsure what the best way to add emoji to them. We could add a swipe option, or we could add the emoji key. Any recommendations? I would love some input from someone :).
```
T9v1.kt // Works fine, just missing the emoji key.
TypeSplitDEv1.kt
TypeSplitENv2.kt

TypeSplitESv1.kt
TypeSplitFRv1.kt
TypeSplitITv1.kt
TypeSplitNumeric.kt
TypeSplitPLv1.kt
TypeSplitPTv1.kt
TypesSplitFIv1.kt
WideLayoutENProgrammer.kt
```

### CODE REVIEW

File: app/src/main/java/com/dessalines/thumbkey/keyboards/CommonKeys.kt
Comment (dessalines) [inline app/src/main/java/com/dessalines/thumbkey/keyboards/CommonKeys.kt:78]: You can remove these two `// Undo` and Redo comments, they're unecessary once you extract them into KeyActions.

File: app/src/main/java/com/dessalines/thumbkey/keyboards/CommonKeys.kt
Comment (dessalines) [inline app/src/main/java/com/dessalines/thumbkey/keyboards/CommonKeys.kt:114]: There's a `FOUR_WAY_CROSS` for this one.

File: app/src/main/java/com/dessalines/thumbkey/keyboards/CommonKeys.kt
Comment (dessalines) [inline app/src/main/java/com/dessalines/thumbkey/keyboards/CommonKeys.kt:115]: I like how you've laid these out.

File: app/src/main/java/com/dessalines/thumbkey/utils/Types.kt
Comment (dessalines) [inline app/src/main/java/com/dessalines/thumbkey/utils/Types.kt:49]: It looks like `SelectAndCopyAll` isn't used anymore, so you can delete it, and the `copy_all` string in strings.xml.

File: app/src/main/java/com/dessalines/thumbkey/utils/Types.kt
Comment (dessalines) [inline app/src/main/java/com/dessalines/thumbkey/utils/Types.kt:51]: Add the Redo and Undo actions here.

File: app/src/main/java/com/dessalines/thumbkey/utils/Utils.kt
Comment (dessalines) [inline app/src/main/java/com/dessalines/thumbkey/utils/Utils.kt:344]: Can probably remove this logging line.

File: app/src/main/java/com/dessalines/thumbkey/utils/Utils.kt
Comment (dessalines) [inline app/src/main/java/com/dessalines/thumbkey/utils/Utils.kt:381]: This entire key action doesn't seem to be used anymore, so you can remove this block.

File: fastlane/metadata/android/en-US/images/phoneScreenshots/1.png
Comment (dessalines) [inline fastlane/metadata/android/en-US/images/phoneScreenshots/1.png:1]: Thx!

File: app/src/main/java/com/dessalines/thumbkey/keyboards/CommonKeys.kt
Comment (dessalines) [inline app/src/main/java/com/dessalines/thumbkey/keyboards/CommonKeys.kt:41]: Why not just get rid of the SETTINGS_KEY_ITEM entirely?

File: app/src/main/java/com/dessalines/thumbkey/ui/components/keyboard/KeyboardScreen.kt
Comment (dessalines) [inline app/src/main/java/com/dessalines/thumbkey/ui/components/keyboard/KeyboardScreen.kt:111]: Move this key definition into `CommonKeys` with the rest of them.

File: app/src/main/java/com/dessalines/thumbkey/ui/components/keyboard/KeyboardScreen.kt
Comment (dessalines) [inline app/src/main/java/com/dessalines/thumbkey/ui/components/keyboard/KeyboardScreen.kt:147]: Add `.background(MaterialTheme.colorScheme.background` here. Otherwise its transparent and doesn't look right on light or dark themes.

File: app/src/main/java/com/dessalines/thumbkey/ui/components/keyboard/KeyboardScreen.kt
Comment (dessalines) [inline app/src/main/java/com/dessalines/thumbkey/ui/components/keyboard/KeyboardScreen.kt:121]: The order of the return and backspace need to be switched.

File: app/src/main/java/com/dessalines/thumbkey/ui/components/keyboard/KeyboardScreen.kt
Comment (dessalines) [inline app/src/main/java/com/dessalines/thumbkey/ui/components/keyboard/KeyboardScreen.kt:110]: It *would* be possible to move everything inside of this block into its own composable, which might be better for organization's sake. 

I leave that up to you, and it can always be done at a later time too.

**Review by dessalines — COMMENTED** [review]
This works great! Thanks a ton for this. Just a few fixes / suggestions.

**Comment by dessalines** [comment]
> In KeyboardScreen.kt I have duplicate code for the massive function call KeyboardKey(). I couldn't figure out a better way to do it.

I don't think there is... it looks verbose, but there's no way around passing those required params.

> I don't use these layouts, so I'm unsure what the best way to add emoji to them. We could add a swipe option, or we could add the emoji key. Any recommendations? I would love some input from someone :).

That's up to those maintainers to upkeep, not you or me. Just replacing the settings key with the new emoji one should be enough for this, and if they'd like to fix / remove some things, they can do so at a later time.

File: app/src/main/java/com/dessalines/thumbkey/keyboards/CommonKeys.kt
Comment (sslater11) [reply app/src/main/java/com/dessalines/thumbkey/keyboards/CommonKeys.kt:80]: I tried that, but it just doesn't work. Nothing happens when I swipe to undo.
I made a `data object Undo : KeyAction()` in Types.kt, and put the SendEvent code into performKeyAction() method in Utils.kt. Not sure why it doesn't work, but it doesn't. The code is executed, it's just the SendEvent that doesn't want to work. I also wanted a nice KeyAction.Undo object to make it easier for keyboard designers.
```
        KeyAction.Undo -> {
            KeyAction.SendEvent(
                KeyEvent(
                    0, 0, KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_Z, 0, KeyEvent.META_CTRL_ON,
                )
            )
        }
```

File: app/src/main/java/com/dessalines/thumbkey/keyboards/CommonKeys.kt
Comment (sslater11) [reply app/src/main/java/com/dessalines/thumbkey/keyboards/CommonKeys.kt:41]: I was going to delete it, but I thought I'd keep it as there's some PRs with layout updates, but you're right, it's never used and the others will be changed to the emoji key anyway :).

File: app/src/main/java/com/dessalines/thumbkey/ui/components/keyboard/KeyboardScreen.kt
Comment (sslater11) [reply app/src/main/java/com/dessalines/thumbkey/ui/components/keyboard/KeyboardScreen.kt:147]: I still can't figure out how to change the text color for the emoji categories, so was thinking of setting the background to light grey and ignoring dark/light themes.
theres a .setBackgroundColor() option to change the background color, but nothing for the text. .foreground will paint a single color over all of the keyboard, hiding everything underneath it, which is useless to us.

Comment (dessalines) [reply app/src/main/java/com/dessalines/thumbkey/ui/components/keyboard/KeyboardScreen.kt:147]: I tried a few things, but overall it doesn't look too bad on dark theme anyway, they use a gray / orange for the foreground text color that looks okay on both light and dark.

File: app/src/main/java/com/dessalines/thumbkey/keyboards/CommonKeys.kt
Comment (dessalines) [reply app/src/main/java/com/dessalines/thumbkey/keyboards/CommonKeys.kt:80]: Hrm... I'll play around with it and see if I can get it to work.

**Comment by sslater11** [comment]
There we go, I've done all the changes you mentioned, although I kept the logging line, because there's one for the numeric mode and shift mode. We should either keep all of them or remove all of them.
I've not worked on moving the duplicate long function call KeyboardKey() to another function, but as you said, that can be done at a later date, so I can look into it :).

**Review by dessalines — APPROVED** [review]
I just tested and it works great, in dark mode also. Thanks a ton for this, ppl are going to really love it.

**Pull request merged**
